brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+        # cov_point_est = np.cov(beta_hat)
+        # current_vec_U_chlsk_l_AR1 = \
+        #     np.linalg.cholesky(cov_point_est + \
+        #     np.eye(n_C) * 1e-6)[l_idx]
+
+        # We add a tiny diagonal element to the point
+        # estimation of covariance, just in case
+        # the user provides data in which
+        # n_V is smaller than n_C
+
+        # (2) start from identity matrix
+
+        # current_vec_U_chlsk_l_AR1 = np.eye(n_C)[l_idx]
+
+        # (3) random initialization
+
+        current_vec_U_chlsk_l_AR1 = np.random.randn(n_l)
+
+        param0 = np.zeros(n_l + 2)
+        param0_fitU = np.zeros(n_l + n_V)
+        param0_fitV = np.zeros(n_V + n_smooth - 1)
+
+        param0_fitU[0:n_l] = current_vec_U_chlsk_l_AR1
+        param0_fitU[n_l:n_l + n_V] = np.sum(
+            residual[0:-1, :] * residual[1:, :], axis=0) / \
+            np.sum(residual[0:-1, :] * residual[0:-1, :], axis=0)
+        log_sigma2_est = np.log(np.var(
+            residual[1:, :] - residual[0:-1, :]
+            * param0_fitU[n_l:n_l + n_V], axis=0))
+        param0_fitU[n_l:n_l + n_V] = \
+            np.tan(param0_fitU[n_l:n_l + n_V] * np.pi / 2)
+        # up to this point, the GP parameters are not set.
+
+        param0[0:n_l] = current_vec_U_chlsk_l_AR1
+        param0[n_l] = np.median(log_sigma2_est)
+        param0[n_l + 1] = np.median(param0_fitU[n_l:n_l + n_V])
+
+        res = scipy.optimize.minimize(
+            self._loglike_y_U_AR1_singpara, param0,
+            args=(XTX, XTDX, XTFX, YTY_diag, YTDY_diag, YTFY_diag,
+                  XTY, XTDY, XTFY, l_idx, n_C, n_T, n_V, rank),
+            method='BFGS', jac=True, tol=1e-2,
+            options={'xtol': self.tolx, 'disp': self.verbose})
+
+        current_a1 = np.ones(n_V) * res.x[n_l + 1]
+        current_logSigma2 = log_sigma2_est
+        # log_sigma2_est is obtained before the above fitting.
+        # Although it should overestimate the variance,
+        # setting it this way would allow it to track log(sigma^2)
+        # more closely for each voxel.
+        current_logSNR2 = -current_logSigma2
+        norm_factor = np.mean(current_logSNR2)
+        current_logSNR2 = current_logSNR2 - norm_factor
+        current_vec_U_chlsk_l_AR1 = res.x[0:n_l] \
+            * np.exp(norm_factor / 2.0)
+
+        param0_fitU[0:n_l] = current_vec_U_chlsk_l_AR1
+        param0_fitU[n_l:n_l + n_V] = current_a1
+        param0_fitV[0:n_V - 1] = current_logSNR2[0:n_V - 1]
+
+        tolx = self.tolx * 5
+        # If each parameter's change is smaller than
+        # this threshold, we stop fitting.
+
+        # We cannot start fitting smoothness parameter just after
+        # fitting the single parameter version. The reason is
+        # that we would have SNR equal in all voxels to start with.
+        # This will favor a very large smoothness length scale.
+        # Therefore, we first fit for some iterations
+        # without the smoothness parameter.
+
+        if GP_space:
```

Comment:
In general this should be multiple functions. Probably the best way is to have functions for estimating U and V in the full model, plus the constant-noise estimate, plus the no-GP estimate. Then the sequence of calls in the main code will make it very clear what the full iterative estimation looks like. Right now it is hard to follow.